jemalloc-ctl/src/raw.rs

```diff
@@ -0,0 +1,244 @@
+    unsafe {
+        cvt(jemalloc_sys::mallctlbymib(
+            mib.as_ptr(),
+            mib.len(),
+            ptr::null_mut(),
+            ptr::null_mut(),
+            &mut value as *mut _ as *mut _,
+            mem::size_of::<T>(),
+        ))
+    }
+}
+
+/// Uses the null-terminated string `name` as the key to the _MALLCTL NAMESPACE_
+/// and sets it `value`
+pub fn set<T>(name: &[u8], mut value: T) -> Result<()> {
+    unsafe {
+        validate_name(name)?;
+
+        cvt(jemalloc_sys::mallctl(
+            name as *const _ as *const c_char,
+            ptr::null_mut(),
+            ptr::null_mut(),
+            &mut value as *mut _ as *mut _,
+            mem::size_of::<T>(),
+        ))
+    }
+}
+
+/// Uses the MIB `mib` as key to the _MALLCTL NAMESPACE_ and sets its `value`
+/// returning its previous value.
+///
+/// The [`name_to_mib`] API translates a string of the key (e.g. `arenas.nbins`)
+/// to a `mib` (Management Information Base).
+pub fn get_set_mib<T>(mib: &[usize], mut value: T) -> Result<T> {
+    unsafe {
+        let mut len = mem::size_of::<T>();
+        cvt(jemalloc_sys::mallctlbymib(
+            mib.as_ptr(),
+            mib.len(),
+            &mut value as *mut _ as *mut _,
+            &mut len,
+            &mut value as *mut _ as *mut _,
+            len,
+        ))?;
+        assert_eq!(len, mem::size_of::<T>());
+        Ok(value)
+    }
+}
+
+/// Uses the null-terminated string `name` as key to the _MALLCTL NAMESPACE_ and
+/// sets its `value` returning its previous value.
+pub fn get_set<T>(name: &[u8], mut value: T) -> Result<T> {
+    unsafe {
+        validate_name(name)?;
+
+        let mut len = mem::size_of::<T>();
+        cvt(jemalloc_sys::mallctl(
+            name as *const _ as *const c_char,
+            &mut value as *mut _ as *mut _,
+            &mut len,
+            &mut value as *mut _ as *mut _,
+            len,
+        ))?;
+        assert_eq!(len, mem::size_of::<T>());
+        Ok(value)
+    }
+}
+
+/// Uses the MIB `mib` as key to the _MALLCTL NAMESPACE_ and reads its value of
+/// type `&str`
+///
+/// The [`name_to_mib`] API translates a string of the key (e.g. `arenas.nbins`)
+/// to a `mib` (Management Information Base).
+pub fn get_str_mib(mib: &[usize]) -> Result<&'static str> {
+    unsafe {
+        let ptr: *const c_char = get_mib(mib)?;
+        ptr2str(ptr)
+    }
+}
+
+/// Uses the null-terminated string `name` as key to the _MALLCTL NAMESPACE_ and
+/// reads its value of type `&str`.
+pub fn get_str(name: &[u8]) -> Result<&'static str> {
+    unsafe {
+        validate_name(name)?;
+
+        let ptr: *const c_char = get(name)?;
+        ptr2str(ptr)
+    }
+}
+
+unsafe fn ptr2str(ptr: *const c_char) -> Result<&'static str> {
+    let len = libc::strlen(ptr);
+    let byte_slice = slice::from_raw_parts(ptr as *const u8, (len + 1) as _);
+    core::str::from_utf8(byte_slice).map_err(|_| Error::EINVAL)
+}
+
+fn validate_name(name: &[u8]) -> Result<()> {
```

Comment:
This should just assert IMO.

Reply:
Yes, I agree, I made it return an error to make the `jemallocator` tests pass, but I'll just refactor those as `#[should_panic]` instead.